fog/report/validation/src/lib.rs

```diff
@@ -57,3 +68,92 @@ pub struct FullyValidatedFogPubkey {
     /// transaction larger than this limit if the fog pubkey is used.
     pub pubkey_expiry: u64,
 }
+
+/// Represents a set of fully validated fog public keys
+///
+/// These can be fetched before building a transaction, then the rest of the
+/// transaction building process can happen offline.
+#[derive(Default, Debug)]
+pub struct FullyValidatedFogPubkeys {
+    /// Map from fog_url, fog_report_id strings to fully validated pubkeys
+    keys: BTreeMap<(String, String), FullyValidatedFogPubkey>,
+}
+
+impl FullyValidatedFogPubkeys {
+    /// Check if we have a fog pubkey needed for a given recipient.
+    /// If so, do nothing.
+    /// If not, try to fetch one using the resolver.
+    pub fn add_key_for_recipient<R: FogPubkeyResolver>(
+        &mut self,
+        resolver: &R,
+        recipient: &PublicAddress,
+    ) -> Result<(), R::Error> {
+        if let Some(table_key) = self.table_key(recipient) {
+            match self.keys.entry(table_key) {
+                Entry::Occupied(_) => {}
+                Entry::Vacant(ent) => {
+                    ent.insert(resolver.get_fog_pubkey(recipient)?);
+                }
+            }
+        }
+        Ok(())
+    }
+
+    /// Alternative to add_key_for_recipient, this may be helpful for SDK code which doesn't want to make blocking calls into rust.
+    pub fn add_key(&mut self, report_url: String, validated_key: FullyValidatedFogPubkey) {
+        self.keys.insert(
+            (report_url, validated_key.fog_report_id.clone()),
+            validated_key,
+        );
+    }
+
+    /// Construct the key to use in the table for given public address
+    /// Returns None if the public address doesn't have a fog report url.
+    fn table_key(&self, addr: &PublicAddress) -> Option<(String, String)> {
+        addr.fog_report_url().map(|url| {
+            (
+                url.to_string(),
+                addr.fog_report_id().unwrap_or("").to_string(),
+            )
+        })
+    }
+
+    /// Get the smallest pubkey_expiry value of any of these fog pubkeys.
+    /// Returns u64::MAX if there are no fog pubkeys present.
+    pub fn smallest_pubkey_expiry_value(&self) -> u64 {
+        self.keys.values().fold(u64::MAX, |cur_min, key| {
+            core::cmp::min(cur_min, key.pubkey_expiry)
+        })
+    }
+}
+
+impl FogPubkeyResolver for FullyValidatedFogPubkeys {
+    type Error = FogPubkeyError;
+
+    fn get_fog_pubkey(
+        &self,
+        recipient: &PublicAddress,
+    ) -> Result<FullyValidatedFogPubkey, Self::Error> {
+        if let Some(table_key) = self.table_key(recipient) {
+            if let Some(result) = self.keys.get(&table_key) {
+                Ok(result.clone())
+            } else {
+                Err(FogPubkeyError::FogPubkeyNotPrefetched(
+                    table_key.0,
+                    table_key.1,
+                ))
+            }
+        } else {
+            Err(FogPubkeyError::NoFogReportUrl)
+        }
+    }
+}
```

Comment:
a problem with this is that we're not validating the fog sig of each user

it might be better to say, this class should hold the reports that passed IAS validation, but when we `get_fog_pubkey` for each respective user, that's when we check the sig

Reply:
idk, the code for checking the sig doesn't exist quite yet, so maybe we can do that later